File: backend/app/ai/llm/rate_limit.py
"""Per-provider client-side rate limiting.

Free-tier LLM quotas are requests-PER-MINUTE, not a fixed total -- bursting past the
limit trips the provider's own 429, which then trips `breaker`'s cooldown and discards
every remaining call in the job (a 24-page answer sheet firing ~24 vision-correction
calls in a few seconds blows straight through a 15 rpm free tier on the first batch).

Pacing calls to stay under the limit means every call gets a real chance to succeed
instead of the whole burst getting shut out after the first one. This module blocks
the calling thread until a slot is free, rather than rejecting the call outright --
`vision.py` fires a handful of pages concurrently, so it must be thread-safe, and
sleeps happen outside the lock so other threads can keep checking their own turn.
"""
from __future__ import annotations

import threading
import time
from collections import deque

_lock = threading.Lock()
_windows: dict[str, deque[float]] = {}

_WINDOW_SECONDS = 60.0


def acquire(provider: str, requests_per_minute: int) -> None:
    """Block until a call for `provider` is safe to make under its RPM cap.

    A cap of 0 or less disables limiting entirely.
    """
    if requests_per_minute <= 0:
        return
    while True:
        with _lock:
            history = _windows.setdefault(provider, deque())
            now = time.monotonic()
            while history and now - history[0] >= _WINDOW_SECONDS:
                history.popleft()
            if len(history) < requests_per_minute:
                history.append(now)
                return
            wait_for = _WINDOW_SECONDS - (now - history[0]) + 0.01
        time.sleep(max(0.0, wait_for))


def reset(provider: str | None = None) -> None:
    """Clear tracked call history. For tests."""
    with _lock:
        if provider is None:
            _windows.clear()
        else:
            _windows.pop(provider, None)

File: apps/web/lib/server/ai/rateLimit.ts
/**
 * Per-model client-side rate limiting. Port of backend/app/ai/llm/rate_limit.py.
 *
 * Free-tier quotas are requests-PER-MINUTE, not a fixed total — bursting past the
 * limit trips the provider's own 429, which then trips the breaker and discards
 * every remaining call. Pacing calls to stay under the limit means every call gets a
 * real chance to succeed instead of the whole burst getting shut out after the first
 * one (e.g. a 24-page answer sheet firing ~24 OCR calls in a few seconds).
 */
const WINDOW_MS = 60_000;
const windows = new Map<string, number[]>();

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Blocks until a call for `key` is safe to make under its RPM cap. 0 disables limiting. */
export async function acquire(key: string, requestsPerMinute: number): Promise<void> {
  if (requestsPerMinute <= 0) return;
  for (;;) {
    const now = Date.now();
    const history = (windows.get(key) ?? []).filter((t) => now - t < WINDOW_MS);
    if (history.length < requestsPerMinute) {
      history.push(now);
      windows.set(key, history);
      return;
    }
    windows.set(key, history);
    const waitFor = WINDOW_MS - (now - history[0]) + 10;
    await sleep(Math.max(0, waitFor));
  }
}

export function reset(key?: string): void {
  if (key) windows.delete(key);
  else windows.clear();
}
